### ISSUE TITLE
Warning with -Wextra

### ISSUE THREAD

**Issue by thibault-martinez** [troydhanson/uthash#176]
Hi,

First, great library! It is being extremely useful for us in [entangled](https://github.com/iotaledger/entangled).

I see in the example [Makefile](https://github.com/troydhanson/uthash/blob/master/tests/Makefile) that `-Wextra` is commented out.
There is a warning in [utarray.h](https://github.com/troydhanson/uthash/blob/master/src/utarray.h#L231) when activated.

```
../src/utarray.h:231:91: warning: signed and unsigned type in conditional expression [-Wsign-compare]
 #define utarray_eltidx(a,e) (((char*)(e) >= (a)->d) ? (((char*)(e) - (a)->d)/(a)->icd.sz) : -1)
```

Are you aware of it and is it planned to fix it ?
We would like to enable `-Werror` in our project.

Thanks in advance,

Thibault,

**Comment by Quuxplusone** [comment]
I'm guessing `-Werror` was a typo for `-Wextra`?
The tests are far from `-Wextra`-clean, which is why it's turned off in the Makefile:

```
test86.c:9:14: warning: unused parameter 'argc' [-Wunused-parameter]
int main(int argc, char *argv[])
             ^
test87.c:42:25: warning: missing field 'prev' initializer [-Wmissing-field-initializers]
        {"muh1", 2,   {0}},
                        ^
```

but I don't see any `-Wsign-compare` warnings in the test suite at the moment. Could you provide a [short example](http://sscce.org) of a program that currently warns? In fact, if you provided your example in the form of a program named `test96.c`, and simultaneously provided a patch for `utarray.h` that silenced the conversion warning, you'd basically have a pull request ready to go. :)

**Comment by thibault-martinez** [comment]
Hi @Quuxplusone , thanks for your answer.

We already use `-Wextra` at the moment and would like to enable `-Werror` but it wouldn't compile due to this warning.

The issue is that both hands of the ternary operation in `utarray_eltidx` doesn't have the same signedness (`icd.sz` which is `size_t` vs `-1` which is `ssize_t`); hence the warning. The `-1` is actually promoted to `size_t` so I think this macro never returns `-1` but its unsigned equivalent.

Can it be problematic since in some other macros you rely on this value being `> 0` ? Or am I missing something ?

Quick example (tested on `Linux c42ac1006624 4.9.125-linuxkit #1 SMP Fri Sep 7 08:20:28 UTC 2018 x86_64 GNU/Linux`)

```
#include <stdio.h>

#include "utarray.h"

int main() {
  UT_array *nums;

  utarray_new(nums, &ut_int_icd);
  for (int i = 0; i < 10; i++) {
    utarray_push_back(nums, &i);
  }

  if (utarray_eltidx(nums, NULL) > 0) {
    fprintf(stderr, "Shouldn't be logged\n");
  }

  utarray_free(nums);

  return 0;
}
```

Do you agree the log shouldn't appear ?

**Comment by Quuxplusone** [comment]
I know [the documentation](http://troydhanson.github.io/uthash/utarray.html) for utarray's API isn't great, but I'm fairly sure that `utarray_eltidx(nums, NULL)` is not expected to do anything sensible. The point of `utarray_eltidx` is to convert from the address of an existing element to that element's index. `NULL` isn't a pointer to an existing element, so trying to use `utarray_eltidx` on it should have undefined behavior, as far as I'm concerned.

However, I've reproduced the warning that I assume you're seeing: https://godbolt.org/z/V4XHXk
```
#include "utarray.h"
int main() {
  UT_array *nums;
  utarray_new(nums, &ut_int_icd);
  int *first = (int*)utarray_next(nums, NULL);
  (void)first;
  utarray_free(nums);
}
```
produces a warning with GCC trunk (interestingly, only in C mode, not in C++ mode)
```
test.c: In function 'main':
utarray.h:231:93: error: operand of '?:' changes signedness from 'int' to 'long unsigned int' due to unsignedness of other operand [-Werror=sign-compare]
utarray.h:228:75: note: in expansion of macro 'utarray_eltidx'
test.c:10:22: note: in expansion of macro 'utarray_next'
```

So I propose this fix: 4b04de41161d0488707cf0a2cc79d237aa852f1a Does that look good to you?

**Comment by thibault-martinez** [comment]
@Quuxplusone Sounds good! Thank you very much :)
